packager/mux/mux_test.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mux
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	pkgt "github.com/ampproject/amppackager/packager/testing"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+// expand propagates hardcoded values into template test url.
+func expand(templateURL string) string {
+	templateURL = strings.Replace(templateURL, "$HOST", "http://www.publisher_amp_server.com", 1)
+	templateURL = strings.Replace(templateURL, "$FETCH", "http://www.publisher_main_server.com/some_page", 1)
+	templateURL = strings.Replace(templateURL, "$SIGN", "https://www.publisher_main_server.com/some_page", 1)
+	templateURL = strings.Replace(templateURL, "$CERT", pkgt.CertName, 1)
+	return templateURL
+}
+
+// mockedHandler mocks mux' underlying http handlers - signer, cert etc.
+type mockedHandler struct {
+	mock.Mock
+}
+
+func (m *mockedHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	m.Called(Params(req))
+}
+
+func TestServeHTTPSuccess(t *testing.T) {
+	templateTests := []struct {
+		testName      string
+		testURL       string
+		expectHandler string
+		expectParams  map[string]string
+	}{
+		{
+			testName:      "Signer - empty",
+			testURL:       "$HOST/priv/doc",
+			expectHandler: "signer",
+			expectParams:  map[string]string{},
+		}, {
+			testName:      "Signer - with query, empty",
+			testURL:       "$HOST/priv/doc?",
+			expectHandler: "signer",
+			expectParams:  map[string]string{},
+		}, {
+			testName:      "Signer - with query, regular",
+			testURL:       "$HOST/priv/doc?fetch=$FETCH&sign=$SIGN",
+			expectHandler: "signer",
+			expectParams:  map[string]string{},
+		},
+		{
+			testName:      "Signer - with query, escaping",
+			testURL:       "$HOST/priv/doc?fetch=$FETCH&sign=$SIGN%2A\\",
```

Comment:
Optionally, use the \`backtick\` string literal syntax to avoid the double backslash.

Reply:
Will do.